### PR TITLE
[fix](mtmv) Fix result wrong when query rewrite by mv if query contains null_unsafe equals expression

### PR DESCRIPTION
**Pull request by seawinde**
## Proposed changes

Fix result wrong when query rewrite by mv if query contains null_unsafe equals expression and the expression both side is slot
table orders data is as following:
```sql
    (null, 1, 'o', 9.5, '2023-12-08', 'a', 'b', 1, 'yy'),
    (1, null, 'o', 10.5, '2023-12-08', 'a', 'b', 1, 'yy'),
    (2, 1, null, 11.5, '2023-12-09', 'a', 'b', 1, 'yy'),
    (3, 1, 'o', null, '2023-12-10', 'a', 'b', 1, 'yy'),
    (3, 1, 'o', 33.5, null, 'a', 'b', 1, 'yy'),
    (4, 2, 'o', 43.2, '2023-12-11', null,'d',2, 'mm'),
    (5, 2, 'o', 56.2, '2023-12-12', 'c',null, 2, 'mi'),
    (5, 2, 'o', 1.2, '2023-12-12', 'c','d', null, 'mi');  
```

such as mv def is 
```sq
select count(*), o_orderstatus, o_comment
            from orders
            group by
            o_orderstatus, o_comment;
```
query is as following:
```sql
           select count(*), o_orderstatus, o_comment
            from orders
            where o_orderstatus = o_orderstatus
            group by
            o_orderstatus, o_comment;
```
after rewrite by materialized view, the result is wrong as following, the row contains null should not appear
```sql
+----------+---------------+-----------+
| count(*) | o_orderstatus | o_comment |
+----------+---------------+-----------+
|        1 | NULL          | yy        |
|        1 | o             | mm        |
|        2 | o             | mi        |
|        4 | o             | yy        |
+----------+---------------+-----------+
```
the pr fix this

### CODE REVIEW

**Comment by seawinde** [comment]
run buildall

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/exploration/mv/mapping/EquivalenceClassMapping.java [hunk 1]
```diff
@@ -30,39 +31,41 @@
...
     /**
      * Generate source equivalence set map to target equivalence set
      */
-    public static EquivalenceClassSetMapping generate(EquivalenceClass source, EquivalenceClass target) {
+    public static EquivalenceClassMapping generate(EquivalenceClass source, EquivalenceClass target) {
 
-        Map<Set<SlotReference>, Set<SlotReference>> equivalenceClassSetMap = new HashMap<>();
-        List<Set<SlotReference>> sourceSets = source.getEquivalenceSetList();
-        List<Set<SlotReference>> targetSets = target.getEquivalenceSetList();
+        Map<List<SlotReference>, List<SlotReference>> equivalenceClassSetMap = new HashMap<>();
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/rules/exploration/mv/mapping/EquivalenceClassMapping.java:52]: why change set to list could fix this problem?

Comment (seawinde) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/rules/exploration/mv/mapping/EquivalenceClassMapping.java:52]: such as the expression `o_orderstatus = o_orderstatus` we should compensate `o_orderstatus = o_orderstatus` on materialized view. 
If we record the slot equal expression in set, only get` {o_orderstatus}`, in `Predicates#compensateEquivalence` we couldn't compensate filter. 
Change set to list. we get  {`o_orderstatus, o_orderstatus}`, then we can compensate filter by `Predicates#compensateEquivalence` .

**Comment by seawinde** [comment]
run buildall

**Review by starocean999 — APPROVED** [review]

**Review by morrySnow — APPROVED** [review]

**Pull request merged**